python/boris_solver.py

```diff
@@ -0,0 +1,209 @@
+        return calculate_omega(bfield, mass=self.mass, charge=self.charge,
+                               energy=0.0)
+
+    def radiation_acceleration(self, pos, vel):
+        """
+        Calculate the acceleration from the radiation reaction force 
+        Form is from Ford & O'Connell equation in 3D
+
+        Args: 
+            pos: 3D position array
+            vel: 3D velocity array
+
+        Returns:
+            Array of accelerations
+        """
+        omega = self.get_omega(pos)
+
+        denom = (1 + self.tau**2 * np.dot(omega, omega))
+        acc = np.zeros(3)
+
+        # Larmor terms
+        acc[0] -= self.tau * (omega[2]**2 + omega[1]**2) * vel[0]
+        acc[0] += self.tau * omega[0] * (omega[2] * vel[2] + omega[1] * vel[1])
+
+        acc[1] -= self.tau * (omega[2]**2 + omega[0]**2) * vel[1]
+        acc[1] += self.tau * omega[1] * (omega[2] * vel[2] + omega[0] * vel[0])
+
+        acc[2] -= self.tau * (omega[0]**2 + omega[1]**2) * vel[2]
+        acc[2] += self.tau * omega[2] * (omega[0] * vel[0] + omega[1] * vel[1])
+        
+        acc /= denom
+        
+        return acc
+    
+    def advance_step(self, time_step, x0, v0):
+        """
+        Advances the equations of motions by a single time step using Boris method
+
+        Args:
+            time_step: The distance forward in time we are stepping
+            x0: Initial position
+            v0: Initial velocity
+
+        Returns the position and velocity at the next step
+        """
+
+        gamma_n = 1 / np.sqrt( 1 - (np.linalg.norm(v0)/c)**2 )    
+        u_n = v0 * gamma_n
+        x_n = x0
+        v_n = v0        
+
+        # Half position step
+        x_nplushalf = x_n + u_n * time_step / (2.0 * gamma_n)
+
+        # Do the first half of the radiation reaction acceleration
+        u_minus = u_n + (time_step / 2.0) * self.radiation_acceleration(x_nplushalf, u_n)
+        gamma_minus = np.sqrt( 1 + (np.linalg.norm(u_n)/c)**2 )
+
+        # Rotation step
+        B_nplushalf = np.zeros(3)
+        # Allows us to use a constant field without calculating if necessary
+        if self.calc_b_field is None:
+            B_nplushalf = self.b_field
+        else:
+            B_nplushalf = self.calc_b_field(x_nplushalf[0], x_nplushalf[1], x_nplushalf[2])
+            
+        theta = self.charge * time_step / (self.mass * gamma_minus) * np.linalg.norm(B_nplushalf)
+        t = np.tan(theta/2.0) * B_nplushalf / np.linalg.norm(B_nplushalf) 
+        u_prime = u_minus + np.cross(u_minus, t)
+        u_plus = u_minus + 2.0 / (1.0 + np.dot(t, t)) * np.cross(u_prime, t)
+
+        # Second half of the radiation reaction acceleration
+        u_nplus1 = u_plus + (time_step / 2.0) * self.radiation_acceleration(x_nplushalf, u_plus)
+
+        # Now update position
+        gamma_nplus1 = np.sqrt( 1 + (np.linalg.norm(u_nplus1)/c)**2 )
+        x_nplus1 = x_nplushalf + u_nplus1 * time_step / (2.0 * gamma_nplus1)
```

Comment:
Similar to a previous comment, if you calculate `v_nplus1` first the update to `x_nplus1` is simpler.